### ISSUE TITLE
Role of artificial intelligence in community contributions

### ISSUE THREAD

**Issue by NickKellett** [sfosc/sfosc#50]
There seem to be legal ambiguities or issues around artificial intelligence-produced code and data, and copyright law. In some jurisdictions only work produced by a creative (human) mind is copyrightable. This appears to be explicitly the case in the US and implicitly in Europe and elsewhere. 

Is this a concern, and if so is the following principle sufficient to overcome this: "All contributors intending to have their work incorporated into a distribution must contribute their work under the same terms as the software license they received it under." 

Or is another principle needed to clarify that community contributors/members must be human, or alternatively that artificial intelligence products that explicitly (and legally) can assign their copyright to a specific individual?

**Comment by Beanow** [comment]
Could you explain why this relates to foss sustainability? This seems to me like an issue with whether a license can be enforced in such a case. The only AI contributions I can think of out in the wild now are possibly bots like https://greenkeeper.io/ proposing version bumps.

**Comment by nothingismagick** [comment]
I think that services like this are about to explode:

https://www.deepcode.ai

And this probably also applies to generative algorithms too. I think it's a
valid point for discussion.

**Comment by NickKellett** [comment]
Yes, I styled my question as more about IP ownership since that's a legal issue today. As @nothingismagick suggested I was thinking of code- and data-generating tools like deepcode and not of code-management utilities like greenkeeper. 

If in some legal jurisdictions computer-generated code is not considered IP-protected due to lack of human involvement, is such a contribution a threat to the SFOSC community? Do its members have to ensure that each contribution includes at least some human involvement in order to avoid potential legal threats to what they produce? If they don't, could that ambiguity or risk discourage people from joining and contributing to it?  What if they don't know how the code was generated - does that have to be spelled out each time somebody makes a pull request?

I think there may be longer-term philosophical risks to a community's sustainability. If a significant (presumably ongoing) code contribution is made by an AI or generative agent setup to do so, it won't in itself be motivated to care about free and open source sustainability: is that a problem?

As well, we are perhaps assuming that a contributor/member is a human (at least doing the git interaction). But terms like "classes of contributor" and "membership" don't explicitly rule out computer agents.  In such a case, how do these principles apply to the computer agent:
- "We must have a strong code of conduct, with clear, fair enforcement mechanisms."
- "Any commercial activity around the software must further the sustainability of the community, and the potential for commercial benefit must be available to all."
- "Rules for membership in the community must be published and adhered to."

Is it up to each community to decide whether its members must be human? If so, do they have to identify computer agents vs humans according to this principle:
- "Membership must have requirements for validation of identity"

What if an SFOSC community doesn't think of these things at first and then it becomes an issue later, could that cause a sustainability crisis for them?

Maybe I'm overthinking these things but we don't know what could be possible in the near future and since we are going back to first principles I think it is worth considering at this point, even if we say "All members/contributors are considered to be human" and punt the question down the road :)

**Comment by Beanow** [comment]
Interesting, if I understand correctly there's several angles you're concerned about there.

In terms of licenses and copyright, it will probably depend on each project's concrete strategy for sustainability how big of a deal this is. If license enforcability is a key component, they will need a CLA equivalent for AI's? Signing over copyrights (if an AI has those in the first place) of it's output to some legal entity like a foundation (#59?) that's bound by the principles and commits to enforcing the license.

As far as code of conduct and rules go, I would argue the enforcement steps may be different but if a bot proves to violate those a project could always decide to require a human pass to manually resolve it, or decide the bot is not suitable for them and revoke access? My guess is this could be handled reactively.

> Any commercial activity around the software must further the sustainability of the community, and the potential for commercial benefit must be available to all.
For this I think we can debate whether any kind of support tools the project uses qualifies for this. Github, Travis, existing code quality as-a-service offerings. Many of these aren't all-in on open source and have commercial products. But they typically don't discriminate between open source projects so at least there's equal opportunity?

The membership aspect seems the most tricky to me. If AI contributions grant you a claim to membership and thus voting rights, it may give some weird power dynamics. Though the risk of "hostile takeovers" I think is inherent to strictly using a democratic model (#5 mentions). Automated contributions to gain large influence is one potential way, but so is a big corporation instructing their employees to obtain membership and vote a certain way.

In favor of these kinds of bots though, they'll probably depend on open source in the sense that it provides them with critically valuable training data. They provide improved code quality free of charge in return. At least this is mutually beneficial. If suitable alternatives exist that have stronger commitments to open source that may be preferable.

**Comment by NickKellett** [comment]
> In terms of licenses and copyright, it will probably depend on each project's concrete strategy for sustainability how big of a deal this is. If license enforcability is a key component, they will need a CLA equivalent for AI's? Signing over copyrights (if an AI has those in the first place) of it's output to some legal entity like a foundation (#59?) that's bound by the principles and commits to enforcing the license.
> 
In some legal jurisdictions it appears that if a human isn't involved at all in the underlying code generation there may not be legal protection. So in such a jurisdiction it may require a human to "get involved" enough to justify IP ownership of the AI-generated code. And then I agree that a CLA is a good practice as it indicates intended ownership and transferring rights to the community to use that code.

> As far as code of conduct and rules go, I would argue the enforcement steps may be different but if a bot proves to violate those a project could always decide to require a human pass to manually resolve it, or decide the bot is not suitable for them and revoke access? My guess is this could be handled reactively.
Agreed, that makes sense - or see below where I suggest explicitly calling out whether membership is human-only. That removes all doubt.

> > Any commercial activity around the software must further the sustainability of the community, and the potential for commercial benefit must be available to all.
> For this I think we can debate whether any kind of support tools the project uses qualifies for this. Github, Travis, existing code quality as-a-service offerings. Many of these aren't all-in on open source and have commercial products. But they typically don't discriminate between open source projects so at least there's equal opportunity?
Again I am not sure if support tooling is an issue at all - i'm thinking more of core code an autonomous agent generates that is part of the core software the community produces. Issues with tooling and support processes won't jeopardize the sustainability of a community, but issues around code ownership and membership could.

> The membership aspect seems the most tricky to me. If AI contributions grant you a claim to membership and thus voting rights, it may give some weird power dynamics. Though the risk of "hostile takeovers" I think is inherent to strictly using a democratic model (#5 mentions). Automated contributions to gain large influence is one potential way, but so is a big corporation instructing their employees to obtain membership and vote a certain way.

Agreed - I suggest to keep things simple it may be best to explicitly say that unless otherwise noted, membership is restricted to humans.
